### Counter

In [2]:
# Импортируем объект Counter из модуля collections
from collections import Counter
# Создаём пустой объект Counter
c = Counter()

In [3]:
c['red'] += 1
print(c)

Counter({'red': 1})


In [4]:
cars = ['red', 'blue', 'black', 'black', 'black', 'red', 'blue', 'red', 'white']
c = Counter(cars)
print(c)


Counter({'red': 3, 'black': 3, 'blue': 2, 'white': 1})


In [6]:
cars_moscow = ['black', 'black', 'white', 'black', 'black', 'white', 'yellow', 'yellow', 'yellow']
cars_spb = ['red', 'black', 'black', 'white', 'white', 'yellow', 'yellow', 'red', 'white']

counter_moscow = Counter(cars_moscow)
counter_spb = Counter(cars_spb)
 
print(counter_moscow)
print(counter_spb)
print(counter_moscow + counter_spb)

Counter({'black': 4, 'yellow': 3, 'white': 2})
Counter({'white': 3, 'red': 2, 'black': 2, 'yellow': 2})
Counter({'black': 6, 'white': 5, 'yellow': 5, 'red': 2})


In [7]:
counter_moscow.subtract(counter_spb)
print(counter_moscow)
# Counter({'black': 2, 'yellow': 1, 'white': -1, 'red': -2})

Counter({'black': 2, 'yellow': 1, 'white': -1, 'red': -2})


In [8]:
counter_moscow = Counter(cars_moscow)
counter_spb = Counter(cars_spb)
 
print(counter_moscow - counter_spb)
# Counter({'black': 2, 'yellow': 1})

Counter({'black': 2, 'yellow': 1})


In [9]:
print(*counter_moscow.elements())

black black black black white white yellow yellow yellow


In [11]:
print(counter_moscow.most_common())
print(counter_moscow.most_common(2))

[('black', 4), ('yellow', 3), ('white', 2)]
[('black', 4), ('yellow', 3)]


In [15]:
len(list(counter_moscow))

3

### defaultdict

In [ ]:
from collections import defaultdict

In [18]:
students = [('Ivanov',1),('Smirnov',4),('Petrov',3),('Kuznetsova',1),
            ('Nikitina',2),('Markov',3),('Pavlov',2)]

groups = defaultdict(list)

for student, group in students:
    groups[group].append(student)
 
print(groups)

defaultdict(<class 'list'>, {1: ['Ivanov', 'Kuznetsova'], 4: ['Smirnov'], 3: ['Petrov', 'Markov'], 2: ['Nikitina', 'Pavlov']})


In [19]:
print(groups[3])
# ['Petrov', 'Markov']

['Petrov', 'Markov']


Если запрашиваемого ключа нет в словаре, KeyError не возникнет. Вместо этого будет напечатан пустой элемент, который создаётся в словаре по умолчанию:

In [20]:
print(groups[2021])
# []

[]


Теперь в словаре groups автоматически появился элемент 2021 с пустым списком внутри, несмотря на то что мы его не создавали:

In [21]:
print(groups)

defaultdict(<class 'list'>, {1: ['Ivanov', 'Kuznetsova'], 4: ['Smirnov'], 3: ['Petrov', 'Markov'], 2: ['Nikitina', 'Pavlov'], 2021: []})


### ORDEREDDICT

В далёкие времена (а точнее, до 2018 года) словари в Python не сохраняли порядок ключей, которые в них добавляли. Попробуйте в Codeboard создать несколько раз словарь с одними и теми же ключами и значениями и напечатать его:

In [35]:
# Напоминаем способ создания словаря через список кортежей
# (ключ, значение)
data = [('Ivan', 19),('Mark', 25),('Andrey', 23),('Maria', 20)]
client_ages = dict(data)
print(client_ages)

{'Ivan': 19, 'Mark': 25, 'Andrey': 23, 'Maria': 20}


In [36]:
# Специальный словарь, который гарантирует сохранение ключей в порядке их добавления, 
# называется OrderedDict:

from collections import OrderedDict
data = [('Ivan', 19),('Mark', 25),('Andrey', 23),('Maria', 20)]
ordered_client_ages = OrderedDict(data)
print(ordered_client_ages)

OrderedDict([('Ivan', 19), ('Mark', 25), ('Andrey', 23), ('Maria', 20)])


Можно, например, отсортировать с помощью функции sorted список кортежей при создании из него OrderedDict, и объекты будут добавлены в порядке сортировки:

In [37]:
data = [('Ivan', 19),('Mark', 25),('Andrey', 23),('Maria', 20)]
# Сортируем по второму значению из кортежа, то есть по возрасту
ordered_client_ages = OrderedDict(sorted(data, key=lambda x: x[1]))
print(ordered_client_ages)

OrderedDict([('Ivan', 19), ('Maria', 20), ('Andrey', 23), ('Mark', 25)])


Хорошие новости! Начиная с версии Python 3.7, гарантируется сохранение ключей в том порядке, в котором они добавлялись в словарь. Однако вам следует помнить о том, что в более старых версиях Python порядок ключей не сохраняется. Это важно для обратной совместимости, то есть для корректной работы программы со старыми версиями интерпретатора. Например, если требуется, чтобы код работал и с версиями Python старше 3.7, и в нём используется очерёдность ключей в словаре, необходимо создавать OrderedDict вместо dict.

Узнать версию Python в коде можно из переменной version из модуля sys.

In [39]:
import sys
print(sys.version)

3.7.4 (default, Aug  9 2019, 18:34:13) [MSC v.1915 64 bit (AMD64)]


### DEQUE

Структура данных deque реализована в модуле collections, поэтому вы сразу получаете доступ к возможностям и стека, и очереди.
Создадим пустой дек (deque). Для этого сначала импортируем эту структуру данных из модуля collections, а затем создадим её пустой экземпляр:

In [40]:
from collections import deque
dq = deque()
print(dq)

deque([])


У deque есть четыре ключевые функции:

- `append` (добавить элемент в конец дека);
- `appendleft` (добавить элемент в начало дека);
- `pop` (удалить и вернуть элемент из конца дека);
- `popleft` (удалить и вернуть элемент из начала дека).
    
Также в очередь возможно добавить сразу несколько элементов из итерируемого объекта в дек. Для этого используют функции `extend` (добавить в конец дека) и `extendleft` (добавить в начало дека)

ДОПОЛНИТЕЛЬНЫЕ ФУНКЦИИ
`reverse` позволяет поменять порядок элементов в очереди на обратный:

In [42]:
dq = deque([1,2,3,4,5])
print(dq)
 
dq.reverse()
print(dq)

deque([1, 2, 3, 4, 5])
deque([5, 4, 3, 2, 1])


`rotate` переносит n заданных элементов из конца очереди в начало:

In [43]:
dq = deque([1,2,3,4,5])
print(dq)

dq.rotate(2)
print(dq)

deque([1, 2, 3, 4, 5])
deque([4, 5, 1, 2, 3])


Элементы можно переносить и из начала в конец:

In [44]:
dq = deque([1,2,3,4,5])
print(dq)
# Отрицательное значение аргумента переносит
# n элементов из начала в конец
dq.rotate(-2)
print(dq)

deque([1, 2, 3, 4, 5])
deque([3, 4, 5, 1, 2])


### Задание 3.2
Дан список из кортежей `temps`. На первом месте в кортеже указан год в виде строки, а на втором — средняя температура января в Петербурге в указанном году.

Напишите функцию `check(temps)`, которая будет выводить словарь, в котором ключи — годы, а значения — показатели температуры. Ключи необходимо отсортировать в порядке убывания соответствующих им температур.

Пример списка temps:
```python
temps = [('2000', -4.4), ('2001', -2.5), ('2002', -4.4), ('2003', -9.5)]
```
Пример вывода:
```
OrderedDict([('2001', -2.5), ('2000', -4.4), ('2002', -4.4), ('2003', -9.5)])
```

In [55]:
def check(temps):
    return dict(sorted(temps,key=lambda x: -x[1]))

temps = [('2000', -4.4), ('2001', -2.5), ('2002', -4.4), ('2003', -9.5)]
check(temps)

{'2001': -2.5, '2000': -4.4, '2002': -4.4, '2003': -9.5}

### Задание 4.3
Напишите функцию `brackets(line)`, которая определяет, является ли последовательность из круглых скобок line правильной.

Примечание 1. Какая последовательность скобок правильная?

Правильной скобочной последовательностью назовём такую последовательность скобок, в которой для каждой открывающей скобки есть последующая соответствующая ей закрывающая скобка. Соответственно, остальные скобочные последовательности назовём неправильными. Пустую строку будем считать правильной последовательностью.

Примечание 2.Для решения этой задачи потребуется использовать стек.

Посимвольно переберите строку. Если встретилась открывающаяся скобка, положите её в стек. Если встретилась закрывающаяся скобка, извлеките скобку из стека.
- Если стек пустой, то есть извлечь скобку нельзя, последовательность неправильная.
- Если строка закончилась и стек стал пустым, последовательность правильная.
- Если в стеке остались скобки, последовательность неправильная.   
    
Пример

```python
print(brackets("(()())"))
# True
print(brackets(""))
# True
print(brackets("(()()))"))
# False
```

In [ ]:
Если стек пустой, то есть извлечь скобку нельзя, последовательность неправильная.
Если строка закончилась и стек стал пустым, последовательность правильная.
Если в стеке остались скобки, последовательность неправильная.
Пример


print(brackets("(()())"))
# True
print(brackets(""))
# True
print(brackets("(()()))"))
# False

In [62]:
from collections import deque

def brackets(line):
    dq = deque()
    for brc in line:
        if brc == '(': 
            dq.append(brc)
        if brc == ')':
            try:
                dq.pop()
            except IndexError:
                return  False
    return len(dq) == 0

print(brackets("(()())"))
# True
print(brackets(""))
# True
print(brackets("(()()))"))
# False

### Задание 4.4
В переменных `center`, `south` и `north` хранятся списки из перечней купленных позиций в трёх торговых точках, расположенных в разных районах города.
Вначале избавьтесь от излишней вложенности: в каждой переменной (center, south, north) должен храниться объединённый список купленных товаров без разбиения по чекам.

Пример: `[['Milk', 'Bread'], ['Meat']] -> ['Milk', 'Bread', 'Meat']`

После этого определите, в каком магазине было куплено больше всего товаров.

In [99]:
center = [['Milk', 'Meat'], ['Meat'],['Meat', 'Bread'],['Meat'],['Meat']]
south = [['Milk', 'Bread'], ['Meat'],['Meat']]
north = [['Milk', 'Bread'], ['Meat'],['Milk', 'Bread'], ['Meat'],['Meat']]

from functools import reduce
center = reduce(lambda x,y: x + y, center)
south = reduce(lambda x,y: x + y, south)
north = reduce(lambda x,y: x + y, north)
sorted([('center',len(center)),('south',len(south)),('north',len(north))],
       key=lambda x: x[1])[-1][0]

'north'

### Задание 4.5
Теперь получите объекты-счётчики (Counter) из каждого полученного в предыдущем задании списка покупок и сохраните их в отдельные переменные (они пригодятся для выполнения следующих задач).
Сколько раз покупали самый редкий товар в магазине north? Запишите ответ в числовой форме.

In [100]:
from collections import Counter

center_cnt = Counter(center)
south_cnt = Counter(south)
north_cnt = Counter(north)

print(north_cnt.most_common()[-1][1])

2


In [101]:
north_cnt.most_common()[-1][1]

2

### Задание 4.6
Выберите товар, который в магазине center покупали чаще, чем в магазине north:

In [103]:
(center_cnt - north_cnt)[]

Counter({'Meat': 2})

### Задание 4.9
Дан список кортежей `ratings` с рейтингами кафе. Кортеж состоит из названия и рейтинга кафе.

Необходимо отсортировать кортеж по убыванию рейтинга. Если рейтинги совпадают, то отсортировать кафе дополнительно по названию в алфавитном порядке.

Получите словарь `cafes` с упорядоченными ключами из отсортированного списка, где ключи — названия кафе, а значения — их рейтинг.




In [105]:
ratings = [('Old York', 3.3), ('New Age', 4.6), ('Old Gold', 3.3), ('General Foods', 4.8),
           ('Belissimo', 4.5), ('CakeAndCoffee', 4.2), ('CakeOClock', 4.2), ('CakeTime', 4.1),
           ('WokToWork', 4.9), ('WokAndRice', 4.9), ('Old Wine Cellar', 3.3), ('Nice Cakes', 3.9)]


In [111]:
from collections import OrderedDict
cafes = OrderedDict(sorted(ratings,key=lambda x: (-x[1], x[0])))
cafes

OrderedDict([('WokAndRice', 4.9),
             ('WokToWork', 4.9),
             ('General Foods', 4.8),
             ('New Age', 4.6),
             ('Belissimo', 4.5),
             ('CakeAndCoffee', 4.2),
             ('CakeOClock', 4.2),
             ('CakeTime', 4.1),
             ('Nice Cakes', 3.9),
             ('Old Gold', 3.3),
             ('Old Wine Cellar', 3.3),
             ('Old York', 3.3)])

### Задание 4.10

Напишите функцию `task_manager(tasks)`, которая принимает список задач `tasks` для нескольких серверов. Каждый элемент списка состоит из кортежа `(<номер задачи>, <название сервера>, <высокий приоритет задачи>)`.

Функция должна создавать словарь и заполнять его задачами по следующему принципу: название сервера — ключ, по которому хранится очередь задач для конкретного сервера.

Если поступает задача без высокого приоритета (последний элемент кортежа — False), добавить номер задачи в конец очереди. Если приоритет высокий, добавить номер в начало.

Для словаря используйте `defaultdict`, для очереди — `deque`.

Функция возвращает полученный словарь с задачами.   
    
Пример
```python
tasks = [(36871, 'office', False),
(40690, 'office', False),
(35364, 'voltage', False),
(41667, 'voltage', True),
(33850, 'office', False)]

print(task_manager(tasks))
# defaultdict(, {'voltage': deque([41667, 35364]),
# 'office': deque([36871, 40690, 33850])})
```

In [116]:
tasks = [(36871, 'office', False),
(40690, 'office', False),
(35364, 'voltage', False),
(41667, 'voltage', True),
(33850, 'office', False)]

In [124]:
from collections import deque
from collections import defaultdict

def task_manager(tasks):
    tm = defaultdict(deque)
    for task_number, server_name, high_priority in tasks:
        if high_priority:
            tm[server_name].appendleft(task_number)
        else:
            tm[server_name].append(task_number)
    return tm

print(task_manager(tasks))

defaultdict(<class 'collections.deque'>, {'office': deque([36871, 40690, 33850]), 'voltage': deque([41667, 35364])})


In [125]:
import numpy as np

In [131]:
np.iinfo(np.int8)

iinfo(min=-128, max=127, dtype=int8)

In [136]:
np.iinfo(np.int64)

iinfo(min=-9223372036854775808, max=9223372036854775807, dtype=int64)

In [141]:
print(*sorted(map(str, set(np.sctypeDict.values()))), sep='\n')

<class 'numpy.bool_'>
<class 'numpy.bytes_'>
<class 'numpy.complex128'>
<class 'numpy.complex128'>
<class 'numpy.complex64'>
<class 'numpy.datetime64'>
<class 'numpy.float16'>
<class 'numpy.float32'>
<class 'numpy.float64'>
<class 'numpy.float64'>
<class 'numpy.int16'>
<class 'numpy.int32'>
<class 'numpy.int32'>
<class 'numpy.int64'>
<class 'numpy.int8'>
<class 'numpy.object_'>
<class 'numpy.str_'>
<class 'numpy.timedelta64'>
<class 'numpy.uint16'>
<class 'numpy.uint32'>
<class 'numpy.uint32'>
<class 'numpy.uint64'>
<class 'numpy.uint8'>
<class 'numpy.void'>


In [142]:
np.uint8(-456)

56

In [144]:
np.finfo(np.float32)

finfo(resolution=1e-06, min=-3.4028235e+38, max=3.4028235e+38, dtype=float32)

In [156]:
aa = np.linspace(-6,21,num=60,endpoint=False)
aa.dtype

dtype('float64')

Вам дан массив mystery
```python
mystery = np.array([[-13586,  15203,  28445, -27117,  -1781, -17182, -18049],
       [ 25936, -30968,  -1297,  -4593,   6451,  15790,   7181],
       [ 13348,  28049,  28655,  -6012,  21762,  25397,   8225],
       [ 13240,   7994,  32592,  20149,  13754,  11795,   -564],
       [-21725,  -8681,  30305,  22260, -17918,  12578,  29943],
       [-16841, -25392, -17278,  11740,   5916,    -47, -32037]],
      dtype=np.int16)
```
Выполните все пукты задания по порядку.

Пожалуйста, не меняйте названия переменных, в которых должны появиться ответы.

- В переменную elem_5_3 сохраните элемент из 5 строки и 3 столбца

- В переменную last сохраните элемент из последней строки последнего столбца

- В переменную line_4 сохраните строку 4

- В переменную col_2 сохраните предпоследний столбец

- Из строк 2-4 (включительно) получите столбцы 3-5 (включительно). Результат сохраните в переменную part

- Сохраните в переменную rev последний столбец в обратном порядке

- Сохраните в переменную trans транспонированный массив



In [157]:
mystery = np.array([[-13586,  15203,  28445, -27117,  -1781, -17182, -18049],
       [ 25936, -30968,  -1297,  -4593,   6451,  15790,   7181],
       [ 13348,  28049,  28655,  -6012,  21762,  25397,   8225],
       [ 13240,   7994,  32592,  20149,  13754,  11795,   -564],
       [-21725,  -8681,  30305,  22260, -17918,  12578,  29943],
       [-16841, -25392, -17278,  11740,   5916,    -47, -32037]],
      dtype=np.int16)

In [ ]:
elem_5_3 = mystery[4,2]
last = mystery[-1,-1]
line_4 = mystery[3]
col_2 = mystery[:,-2]
part = mystery[1:4, 2:5]
rev = mystery[:,-1][::-1]
trans = mystery.T

Вам дан массив mystery:
```python
mystery = np.array([ 12279., -26024.,  28745.,  np.nan,  31244.,  -2365.,  -6974.,
        -9212., np.nan, -17722.,  16132.,  25933.,  np.nan, -16431.,
        29810.], dtype=np.float32)
```
- Получите булевый массив `nans_index` с информацией о np.nan в массиве mystery: True - значение пропущено, False - значение не пропущено

- В переменную `n_nan` сохраните число пропущенных значений

- Скопируйте массив `mystery` в массив `mystery_new`. Заполните пропущенные значения в массиве `mystery_new` нулями

- Поменяйте тип данных в массиве `mystery` на `int32` и сохраните в переменную `mystery_int`

- Отсортируйте значения в массиве по возрастанию и сохраните результат в переменную array

- Сохраните в массив `table` двухмерный массив, полученный из массива `array`. В нём должно быть 5 строк и 3 столбца. Причём порядок заполнения должен быть по столбцам!   
Например,
```
 1, 2, 3, 4 -> 1    3
                2    4
```
- Сохраните в переменную `col` средний столбец из `table`   

Примечание. Не меняйте названия переменных.



In [201]:
import numpy as np

mystery = np.array([ 12279., -26024.,  28745.,  np.nan,  31244.,  -2365.,  -6974.,
        -9212., np.nan, -17722.,  16132.,  25933.,  np.nan, -16431.,
        29810.], dtype=np.float32)

nans_index = np.isnan(mystery)

n_nan = nans_index.sum()

mystery_new = mystery.copy()
mystery_new[nans_index] = 0

mystery_int = mystery.astype(np.int32)

array = np.sort(mystery)

table = array.reshape((5,3),order='F')

col = table[:, 1]

In [ ]:
mport numpy as np
np.random.seed(2021)
simple = np.random.rand()
randoms = np.random.uniform(-150, 2021, size=120)
table = np.random.randint(1, 101, size=(3,2))
even = np.arange(2,17,2)
mix = even
np.random.shuffle(mix)
select = np.random.choice(even, replace=False, size=3)
triplet = np.random.permutation(select)

In [204]:
np.iinfo(np.int16)

iinfo(min=-32768, max=32767, dtype=int16)

In [ ]:
import numpy as np

def shuffle_seed(array):
    seed = np.random.randint(2 ** 32)
    np.random.seed(seed)
    result = np.random.permutation(array)
    return result, seed

In [227]:
np.random.randint(2)

1

In [228]:
import numpy as np

def get_unique_loto(num):
    sample = np.arange(1, 101)
    res = list()
    for i in range(num):
        res.append(np.random.choice(sample, replace=False, size=(5, 5)))
    res = np.array(res)
    return res

In [229]:
get_unique_loto(2)

array([[[ 69,   3,  61,   4,  90],
        [ 40,  54,  96,  38,  65],
        [ 24,  27,  60, 100,  59],
        [ 19,  12,  48,  72,  70],
        [ 20,  41,  91,  25,  16]],

       [[ 19,  94,   8,  32,  11],
        [ 84,  46,  37,  77,   3],
        [  7,  22,  98,  16,  56],
        [ 88,  55,  71,  15,  35],
        [ 60,  83,  78,   5,  58]]])

In [233]:
simplelist = [19, 242, 14, 152, 142, 1000]
np.mean(simplelist)


261.5